app/src/main/kotlin/compiler/semantic_analysis/ArgumentResolver.kt

```diff
@@ -2,17 +2,182 @@ package compiler.semantic_analysis
+    private val result = ReferenceHashMap<Expression.FunctionCall.Argument, Function.Parameter>()
+    private var correctDefinitions = true
+
+    companion object {
+        fun calculateArgumentToParameterResolution(
+            program: Program,
+            nameResolution: ReferenceMap<Any, NamedNode>,
+            diagnostics: Diagnostics
+        ): ArgumentResolutionResult {
+            val resolver = ArgumentResolver(nameResolution, diagnostics)
+
+            resolver.checkFunctionDefinitions(program)
+            if (resolver.correctDefinitions)
+                resolver.checkFunctionCalls(program)
+
+            return resolver.result
+        }
+    }
+
+    private fun checkFunctionCall(functionCall: Expression.FunctionCall) {
+        var foundNamedArgument = false
+        for (argument in functionCall.arguments) {
+            if (argument.name != null) {
+                foundNamedArgument = true
+            } else if (foundNamedArgument) {
+                diagnostics.report(Diagnostic.ArgumentResolutionError.PositionalArgumentAfterNamed(functionCall))
+                return
+            }
+        }
+
+        val function: Function = nameResolution[functionCall] as Function
+        val parameterNames = function.parameters.map { it.name }
+        val numberOfParameters = parameterNames.size
+        val isMatched = parameterNames.map { false }.toMutableList()
+
+        for ((index, argument) in functionCall.arguments.withIndex()) {
+            if (argument.name == null) {
+                if (index >= numberOfParameters) {
+                    diagnostics.report(Diagnostic.ArgumentResolutionError.TooManyArguments(functionCall))
+                    return
+                }
+                result[argument] = function.parameters[index]
+                isMatched[index] = true
+            } else {
+                val parameterIndex = parameterNames.indexOf(argument.name)
+                if (parameterIndex == -1) {
+                    diagnostics.report(Diagnostic.ArgumentResolutionError.UnknownArgument(functionCall, argument.name))
+                    return
+                }
+                if (isMatched[parameterIndex]) {
+                    diagnostics.report(Diagnostic.ArgumentResolutionError.RepeatedArgument(functionCall, argument.name))
+                    return
+                }
+                result[argument] = function.parameters[parameterIndex]
+                isMatched[parameterIndex] = true
+            }
+        }
+
+        for (index in 0 until numberOfParameters) {
+            if (!isMatched[index] && function.parameters[index].defaultValue == null) {
+                diagnostics.report(Diagnostic.ArgumentResolutionError.MissingArgument(functionCall, parameterNames[index]))
+            }
+
+            // TODO: detect default parameter values used in this function call
```

Comment:
is this still relevant?

Reply:
You're right about `functionCall.arguments`. I feel like mapping from function calls to default paramters used in that call is needed in later analysis (it uses variables which store default parameter values for local functions).

Reply:
Ok, so we can't recur yet, before this `TODO` is completed?

Reply:
We can. I added it to `processExpression`.

Reply:
didn't see, sorry